### ISSUE TITLE
Dynamic SELECT queries

### ISSUE THREAD

**Issue by Bas950** [porsager/postgres#437]
```typescript
async function select(columns: (keyof TSchema)[] | "*") {
	const res = await this.sql<(TSchema & Row)[]>`SELECT ${
		typeof columns === "string"
			? this.sql(columns)
			: this.sql(columns.join(","))
	} FROM ${this.sql(this.schema)}.${this.sql(this.table)}`;
	return res;
}
```

Working:
```typescript
select('*')
```

Not working:
```typescript
select(["id", "column2"])
```
Error:
`PostgresError: column "id,column2" does not exist`

Basically:
`this.sql(columns.join(","))` doesn't work, is there something else that I should be using?

**Comment by Bas950** [comment]
Oh bruh just noticed if you do
```typescript
const res = await this.sql<(TSchema & Row)[]>`SELECT ${this.sql(
	columns as string | string[]
)} FROM ${this.sql(this.schema)}.${this.sql(this.table)}`;
```
it works... so there is some typing issue when you use keyof values...

Typing error without the `as string | string[]`:
```
No overload matches this call.
  Overload 1 of 2, '(first: ((keyof TSchema)[] | "*") & (string | ((keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParameter<...>[]))): Helper<...>', gave the following error.
    Argument of type '(keyof TSchema)[] | "*"' is not assignable to parameter of type '((keyof TSchema)[] | "*") & (string | ((keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParame...'.
      Type '(keyof TSchema)[]' is not assignable to type '((keyof TSchema)[] | "*") & (string | ((keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParame...'.
        Type '(keyof TSchema)[]' is not assignable to type '(keyof TSchema)[] & ((keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParameter<...>[])'.
          Type '(keyof TSchema)[]' is not assignable to type '(keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<never> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParameter<...>[]'.
  Overload 2 of 2, '(template: TemplateStringsArray, ...parameters: readonly (PendingQuery<any> | SerializableParameter<never>)[]): PendingQuery<Row[]>', gave the following error.
    Argument of type '(keyof TSchema)[] | "*"' is not assignable to parameter of type 'TemplateStringsArray'.
      Property 'raw' is missing in type '(keyof TSchema)[]' but required in type 'TemplateStringsArray'.
```

**Comment by Minigugus** [comment]
> it works... so there is some typing issue when you use keyof values...

TLDR: add `string &` before `keyof` in `columns` definition, e.g. `const columns: (string & keyof TSchema)[] | '*' = /* your code... */`

`keyof TSchema` returns a type assignable to `string | number | symbol`, but here `postgres` only accepts `string` (that's because `Object.keys(...)` can return symbols and numbers too, not just strings)

**Comment by Bas950** [comment]
> > it works... so there is some typing issue when you use keyof values...
> 
> TLDR: add `string &` before `keyof` in `columns` definition, e.g. `const columns: (string & keyof TSchema)[] | '*' = /* your code... */`
> 
> `keyof TSchema` returns a type assignable to `string | number | symbol`, but here `postgres` only accepts `string` (that's because `Object.keys(...)` can return symbols and numbers too, not just strings)

```
No overload matches this call.
  Overload 1 of 2, '(first: ((string & keyof TSchema)[] | "*") & (string | ((string & keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (string & keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParameter<...>[]))): Helper<...>', gave the following error.
    Argument of type '(string & keyof TSchema)[] | "*"' is not assignable to parameter of type '((string & keyof TSchema)[] | "*") & (string | ((string & keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (string & keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : ...'.
      Type '(string & keyof TSchema)[]' is not assignable to type '((string & keyof TSchema)[] | "*") & (string | ((string & keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (string & keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : ...'.
        Type '(string & keyof TSchema)[]' is not assignable to type '(string & keyof TSchema)[] & ((string & keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (string & keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly Serializa...'.
          Type '(string & keyof TSchema)[]' is not assignable to type '(string & keyof TSchema)[] extends readonly any[][] ? readonly EscapableArray[] : (string & keyof TSchema)[] extends readonly (object & infer R)[] ? R extends SerializableParameter<...> ? readonly SerializableParameter<...>[] : readonly (Record<...> & Record<...>)[] : readonly SerializableParameter<...>[]'.
  Overload 2 of 2, '(template: TemplateStringsArray, ...parameters: readonly (PendingQuery<any> | SerializableParameter<never>)[]): PendingQuery<Row[]>', gave the following error.
    Argument of type '(string & keyof TSchema)[] | "*"' is not assignable to parameter of type 'TemplateStringsArray'.
      Property 'raw' is missing in type '(string & keyof TSchema)[]' but required in type 'TemplateStringsArray'.
```

**Comment by Bas950** [comment]
Will close this issue, since it's now just typing related, no longer to postgres itself.
